src/bidiServer/WebSocketServer.ts

```diff
@@ -126,44 +176,52 @@ export class WebSocketServer {
-      debugInternal('new WS request received:', request.resourceURL.path);
+      // Session is set either by Classic or BiDi commands.
+      let session: Session | undefined;
 
-      const connection = request.accept();
+      const requestSessionId = (request.resource ?? '').split('/').pop();
+      debugInternal(
+        `new WS request received. Path: ${JSON.stringify(
+          request.resourceURL.path
+        )}, sessionId: ${JSON.stringify(requestSessionId)}`
+      );
 
-      const browserInstanceDeferred = new Deferred<BrowserInstance>();
+      if (
+        requestSessionId !== '' &&
+        requestSessionId !== undefined &&
+        !this.#sessions.has(requestSessionId)
+      ) {
+        debugInternal('Unknown session id:', requestSessionId);
+        request.reject();
+        return;
+      }
 
-      // Schedule browser instance creation, but don't wait for it.
-      void (async () => {
-        try {
-          debugInfo('Scheduling browser launch...');
-          const browserInstance = await BrowserInstance.run(
-            channel,
-            headless,
-            verbose,
-            chromeOptions?.args
-          );
+      const connection = request.accept();
 
-          // Forward messages from BiDi Mapper to the client unconditionally.
-          browserInstance.bidiSession().on('message', (message) => {
-            void this.#sendClientMessageString(message, connection);
+      session = this.#sessions.get(requestSessionId ?? '');
+      if (session !== undefined) {
+        // BrowserInstance is created for each new WS connection, even for the
+        // same SessionId. This is because WPT uses a single session for all the
+        // tests, but cleans up tests using WebDriver Classic commands, which is
+        // not implemented in this Mapper runner.
+        // TODO: connect to an existing BrowserInstance instead.
+        const sessionOptions = session.sessionOptions;
+        session.browserInstancePromise = this.#closeBrowserInstanceIfLaunched(
+          session
+        )
+          .then(
+            async () =>
+              await this.#launchBrowserInstance(connection, sessionOptions)
+          )
+          .catch((e) => {
+            debugInfo('Error while creating session', e);
+            connection.close(500, 'cannot create browser instance');
+            throw e;
           });
-
-          debugInfo('Browser is launched!');
-          browserInstanceDeferred.resolve(browserInstance);
-        } catch (e) {
-          debugInfo('Error while creating browser instance', e);
-          connection.close(500, 'Error while creating browser instance');
-          return;
-        }
-      })();
+      }
 
       connection.on('message', async (message) => {
-        // Wait for browser instance to be created.
-        const browserInstance = await browserInstanceDeferred;
-
-        // If |type| is not text, return a error.
+        // If |type| is not text, return error.
```

Comment:
```suggestion
        // If type is not text, return error.
```

`||` is spec formatting, not relevant here